externals/llvm-external-projects/torch-mlir-dialects/lib/Conversion/TcpToLinalg/Elementwise.cpp

```diff
@@ -174,24 +174,97 @@ createLinalgPayloadForElementwiseOp(Operation *op,
       return {b.create<arith::MulIOp>(loc, payloadArgs[0], payloadArgs[1])};
     else
-      llvm_unreachable(
-          "unsupported element type in createLinalgPayloadForElementwiseOp");
+      llvm_unreachable("unsupported element type in "
+                       "createLinalgPayloadForElementwiseOp for tcp.mul");
   }
 
   if (isa<DivFOp>(op)) {
     if (elemType.isa<mlir::FloatType>())
       return {b.create<arith::DivFOp>(loc, payloadArgs[0], payloadArgs[1])};
     else
-      llvm_unreachable(
-          "unsupported element type in createLinalgPayloadForElementwiseOp");
+      llvm_unreachable("unsupported element type in "
+                       "createLinalgPayloadForElementwiseOp for tcp.divf");
   }
 
   if (isa<Atan2Op>(op)) {
     if (elemType.isa<mlir::FloatType>())
       return {b.create<math::Atan2Op>(loc, payloadArgs[0], payloadArgs[1])};
     else
-      llvm_unreachable(
-          "unsupported element type in createLinalgPayloadForElementwiseOp");
+      llvm_unreachable("unsupported element type in "
+                       "createLinalgPayloadForElementwiseOp for tcp.atan2");
+  }
+
+  if (auto castOp = dyn_cast<CastOp>(op)) {
+    auto inputType =
+        castOp.getIn().getType().dyn_cast<RankedTensorType>().getElementType();
+    auto outputType = resultTensorType.getElementType();
+
+    if (inputType.getIntOrFloatBitWidth() ==
+            outputType.getIntOrFloatBitWidth() &&
+        ((!castOp.getInIntSignedness() && !castOp.getOutIntSignedness()) ||
+         (castOp.getInIntSignedness() && castOp.getOutIntSignedness() &&
+          castOp.getInIntSignedness().value() ==
+              castOp.getOutIntSignedness().value())))
+      // check for same type
+      return {payloadArgs[0]};
+    else if (outputType.isInteger(1)) {
+      // To I1 (Bool) type
+      Value cstZero =
+          b.create<arith::ConstantOp>(loc, b.getZeroAttr(inputType));
+      if (inputType.isa<mlir::FloatType>()) {
+        return {b.create<arith::CmpFOp>(loc, arith::CmpFPredicate::UNE,
+                                        payloadArgs[0], cstZero)};
+      } else if (inputType.isa<mlir::IntegerType>()) {
+        return {b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ne,
+                                        payloadArgs[0], cstZero)};
+      }
+    } else if (outputType.isa<mlir::FloatType>()) {
+      // TO FP type
+      // FP -> FP
+      if (inputType.dyn_cast<mlir::FloatType>()) {
+        if (inputType.getIntOrFloatBitWidth() >
+            outputType.getIntOrFloatBitWidth())
+          return {b.create<arith::TruncFOp>(loc, outputType, payloadArgs[0])};
+        return {b.create<arith::ExtFOp>(loc, outputType, payloadArgs[0])};
+      }
+      // INT -> FP
+      else if (inputType.dyn_cast<mlir::IntegerType>()) {
+        // Signless or Unsigned INT to FP
+        if (castOp.getInIntSignedness().value() == Signedness::Signless ||
+            castOp.getInIntSignedness().value() == Signedness::Unsigned)
```

Comment:
Why does signless mean unsigned? This was a bit unexpected.

Reply:
IIRC, this bit is meant for `i1` (`bool`) case only. Do you see any issues with that?

@zezhang Can you document this in the op? Also, it would be good to add a check in the verifier to ensure that it is `signless` only when the integer width is 1.

Reply:
"IIRC, this bit is meant for i1 (bool) case only. Do you see any issues with that?". Ah, this was non-obvious :) Additional documentation would help with this.